<!DOCTYPE html>
<!--
To change this license header, choose License Headers in Project Properties.
To change this template file, choose Tools | Templates
and open the template in the editor.
-->
<html xmlns="http://www.w3.org/1999/xhtml"
      xmlns:th="http://www.thymeleaf.org">
    <head>
        <title>Exomiser</title>
        <meta charset="UTF-8"/>
        <meta name="viewport" content="width=device-width, initial-scale=1.0"/>
        <link rel="stylesheet" type="text/css" media="all" href="../../resources/bootstrap-3.2.0/css/bootstrap.min.css" th:href="@{resources/bootstrap-3.2.0/css/bootstrap.min.css}"/>
    </head>
    <body>
        <div class="container">
            <div class="panel panel-default">
                <div class="panel-heading">
                    <h1>The Exome Wobbler: A Tool to Annotate and Filter Variants</h1>
                </div>
                <div class="panel-body">
                    <p class="lead">
                        The Exomiser is a Java program that functionally annotates variants from whole-exome
                        sequencing data starting from a VCF file (version 4). The functional annotation code is
                        based on <a href="http://www.openbioinformatics.org/annovar">Annovar</a> and uses
                        <a href="http://genome.ucsc.edu/">UCSC</a> KnownGene transcript definitions and
                        hg19 genomic coordinates
                    </p>  
                    <p> 
                        Variants are prioritized according to user-defined criteria on variant frequency, pathogenicity,
                        quality, inheritance pattern, and model organism phenotype data. Predicted pathogenicity data
                        was extracted from the <a href="http://www.ncbi.nlm.nih.gov/pubmed/21520341">dbNSFP</a> resource.
                        Cross-species phenotype comparisons come from our <a href="http://www.sanger.ac.uk/resources/databases/phenodigm">PhenoDigm</a> tool powered 
                        by the <a href="http://code.google.com/p/owltools/wiki/OwlSim">OWLSim</a> algorithm.
                    </p>
                    <p>
                        The Exomiser was developed by the Computational Biology and Bioinformatics group at the
                        <a href="http://genetik.charite.de">Institute for Medical Genetics and Human Genetics</a> of the
                        <a href="www.charite.de">Charit&eacute; - Universit&auml;tsmedizin Berlin</a>, the Mouse
                        Informatics Group at the <a href="http://www.sanger.ac.uk/">Sanger Institute</a> and other members of 
                        the <a href="http://monarchinitiative.org/">Monarch initiative</a>.
                    </p>
                    <p>
                        <strong>Read more in our paper: </strong>Robinson P, KÃ¶hler S, Oellrich A, Sanger Mouse Genetics Project, Wang K, Mungall C, Lewis SE, Washington N, Bauer S, Seelow D, Krawitz P, Gilissen C, Haendel M and Smedley D. Improved exome prioritization of disease genes through cross species phenotype comparison. <a href="http://genome.cshlp.org/content/24/2/340.long">Genome Research. 2014 Feb;24(2):340-8.</a>
                    </p>
                    <p>
                        <a href="ftp://ftp.sanger.ac.uk/pub/resources/software/exomiser/downloads/exomiser">Download Exomiser</a> <span class="label label-danger">New</span>
                    </p>

                    <form action="submit" method="GET">
                        <button type="submit" class="btn btn-primary" id="button_go_to_submission">
                            <span class="glyphicon glyphicon-arrow-right"></span> Go to data submission and analysis
                        </button>                            
                    </form>  
                </div>
            </div>
        </div>
    </body>
</html>
